### PR TITLE
Add ci publish workflow

### PR DESCRIPTION
**Pull request by ScottGibb**
# Proposal for CI Publishing to crates.io

The following is a very simple solution to ticket https://github.com/embassy-rs/bt-hci/issues/20. It does not involve any changes to the existing CI infrastructure.

## How it Works
Whenever a new tag is made in the format VX.X.X, the release workflow will activate. It consists of two jobs.
- `check`: Which is identical to the tests,yml. Ensuring that every release runs our ci tests
- `release`: This job will only run if the check job passes, It will first verify that the Github tag matches the config.toml tag. Only then will it continue on with the build process and then publish both a GitHub Release and crates.io release. All tags should match each other, allowing a single source of truth to be found.

## What Other options did I try

- I tried to get the workflow to depend on the tests.yml workflow but couldn't find a good way of doing it.
- I did look into automated PR publishing, however launching on tag means we can choose when we deliver a next release in a low effort fashion. It does rely that main is always tested and in a good stage, otherwise the release workflow will fail (on purpose) preventing a release to crates.io 

## Testing

- Since this requires access to the repos secrets, im unable to see this on my side. I did however make a repo to try out this process, it can be found here: https://github.com/ScottGibb/ci_playground. ~~It does sucessfully publish crates to the staging crates.io https://staging.crates.io/crates/ci_playground~~. It was tested both in CI workflow and on terminal using the `--dry-run` command.

### CODE REVIEW

**Comment by ScottGibb** [comment]
The way in which ive orchestrated this release cycle is under the assumption. That not every PR into main will result in a crates.io release, but rather every tag in the format of vX.X.X on the main branch will result in a new crates.io/GitHub Release

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        if [ "$TAG_VERSION" != "$CARGO_VERSION" ]; then
+          echo "Version mismatch: tag is $TAG_VERSION but Cargo.toml is $CARGO_VERSION"
+          exit 1  # Exits with a non-zero status to fail the workflow
+        fi
+      shell: bash
+
+    - name: Set up Rust
+      uses: dtolnay/rust-toolchain@stable
+      with:
+        toolchain: stable
```
Comment (ScottGibb) [inline .github/workflows/release.yml:47]: Do we want this set to nightly @jamessizeland @lulf?

Comment (petekubiak) [reply .github/workflows/release.yml:47]: Just going to stick my oar in and say that if nightly isn't required I wouldn't use it

**Review by petekubiak — APPROVED** [review]

File: .github/workflows/release.yml (hunk 1 above)
Comment (lulf) [reply .github/workflows/release.yml:47]: Agreed, no need to use nightly

File: README.md
Comment (alexmoon) [inline README.md:1]: I really don't think we need a badge for the publish workflow. Let's revert this file back to its previous state.

File: Cargo.toml [hunk 2]
```diff
@@ -10,7 +10,14 @@ documentation = "https://docs.rs/bt-hci"
...
 rust-version = "1.77"
-exclude = [".github", ".vscode", "ci.sh", "rust-toolchain.toml", "rustfmt.toml", "update_uuids"]
+exclude = [
+    ".github",
+    ".vscode",
+    "ci.sh",
+    "rust-toolchain.toml",
+    "rustfmt.toml",
+    "update_uuids",
+]
```
Comment (alexmoon) [inline Cargo.toml:20]: Please revert this formatting change.

File: .cargo/config.toml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+[registries]
+staging = { index = "https://github.com/rust-lang/staging.crates.io-index" }
```
Comment (alexmoon) [inline .cargo/config.toml:2]: The staging registry isn't for end user testing so we shouldn't be using it here.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    - name: Publish to crates.io
+      env:
+        CARGO_REGISTRIES_STAGING_TOKEN: ${{ secrets.CARGO_REGISTRIES_STAGING_TOKEN }}
+        CARGO_REGISTRY_TOKEN: ${{ secrets.CARGO_REGISTRY_TOKEN }}
+      # Publish to the main registry
+      run: |
+        echo "Publishing to crates.io"  
+        cargo publish
+      # To publish to the staging/testing registry, uncomment the following lines
+      # run: |
```
[the remark this answers is not in the document]
Comment (alexmoon) [reply .github/workflows/release.yml:71]: For testing we should use the `--dry-run` flag instead of the staging registry.

File: Cargo.toml (hunk 2 above)
Comment (ScottGibb) [reply Cargo.toml]: That should be it changed back now :). What is the project using for formatting? As my Rust Analyzer in vscode seems to want keep changing the formatting of that file?

**Comment by ScottGibb** [comment]
If this is suitable to merge @lulf @alexmoon, Could you please squash the commits before merge? Due to the nature of how I tested this solution in GitHub, theres more commits than is really needed for a 1 file change :)

**Review by alexmoon — APPROVED** [review]

**Comment by alexmoon** [comment]
@lulf can you add the CARGO_REGISTRY_TOKEN secret to the github settings so this will work?

**Comment by Dirbaio** [comment]
@lulf try now

**Comment by lulf** [comment]
@alexmoon Done

**Pull request merged**